From 8a88daf6866c8c1d64b118f20f3536d1c1d3d3f7 Mon Sep 17 00:00:00 2001
From: Shaun2029 <shauns2029@gmail.com>
Date: Sat, 12 May 2018 00:37:19 +0100
Subject: [PATCH] EDID: Replaced generic edid/1920x1080.bin EDID with 1920x1080
 256 byte EDID that forces YCbCr colour space. Fixes colours on devices that
 do not correctly support RGB.

---
 drivers/gpu/drm/drm_edid_load.c | 132 ++++++++++++++++++++++++++++----
 1 file changed, 117 insertions(+), 15 deletions(-)

diff --git a/drivers/gpu/drm/drm_edid_load.c b/drivers/gpu/drm/drm_edid_load.c
index a4915099aaa9..2d3226ed349a 100644
--- a/drivers/gpu/drm/drm_edid_load.c
+++ b/drivers/gpu/drm/drm_edid_load.c
@@ -57,7 +57,7 @@ static const char * const generic_edid_name[GENERIC_EDIDS] = {
 	"edid/1920x1080.bin",
 };
 
-static const u8 generic_edid[GENERIC_EDIDS][128] = {
+static const u8 generic_edid[GENERIC_EDIDS][256] = {
 	{
 	0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00,
 	0x31, 0xd8, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
@@ -75,6 +75,23 @@ static const u8 generic_edid[GENERIC_EDIDS][128] = {
 	0x20, 0x20, 0x20, 0x20, 0x00, 0x00, 0x00, 0xfc,
 	0x00, 0x4c, 0x69, 0x6e, 0x75, 0x78, 0x20, 0x53,
 	0x56, 0x47, 0x41, 0x0a, 0x20, 0x20, 0x00, 0xc2,
+
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 	},
 	{
 	0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00,
@@ -93,6 +110,23 @@ static const u8 generic_edid[GENERIC_EDIDS][128] = {
 	0x20, 0x20, 0x20, 0x20, 0x00, 0x00, 0x00, 0xfc,
 	0x00, 0x4c, 0x69, 0x6e, 0x75, 0x78, 0x20, 0x58,
 	0x47, 0x41, 0x0a, 0x20, 0x20, 0x20, 0x00, 0x55,
+
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 	},
 	{
 	0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00,
@@ -111,6 +145,23 @@ static const u8 generic_edid[GENERIC_EDIDS][128] = {
 	0x20, 0x20, 0x20, 0x20, 0x00, 0x00, 0x00, 0xfc,
 	0x00, 0x4c, 0x69, 0x6e, 0x75, 0x78, 0x20, 0x53,
 	0x58, 0x47, 0x41, 0x0a, 0x20, 0x20, 0x00, 0xa0,
+
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 	},
 	{
 	0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00,
@@ -129,6 +180,23 @@ static const u8 generic_edid[GENERIC_EDIDS][128] = {
 	0x20, 0x20, 0x20, 0x20, 0x00, 0x00, 0x00, 0xfc,
 	0x00, 0x4c, 0x69, 0x6e, 0x75, 0x78, 0x20, 0x55,
 	0x58, 0x47, 0x41, 0x0a, 0x20, 0x20, 0x00, 0x9d,
+
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 	},
 	{
 	0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00,
@@ -147,24 +215,58 @@ static const u8 generic_edid[GENERIC_EDIDS][128] = {
 	0x20, 0x20, 0x20, 0x20, 0x00, 0x00, 0x00, 0xfc,
 	0x00, 0x4c, 0x69, 0x6e, 0x75, 0x78, 0x20, 0x57,
 	0x53, 0x58, 0x47, 0x41, 0x0a, 0x20, 0x00, 0x26,
+
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 	},
 	{
 	0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00,
-	0x31, 0xd8, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-	0x05, 0x16, 0x01, 0x03, 0x6d, 0x32, 0x1c, 0x78,
-	0xea, 0x5e, 0xc0, 0xa4, 0x59, 0x4a, 0x98, 0x25,
-	0x20, 0x50, 0x54, 0x00, 0x00, 0x00, 0xd1, 0xc0,
+	0x58, 0xb3, 0x00, 0x37, 0x00, 0x00, 0x00, 0x00,
+	0x01, 0x18, 0x01, 0x03, 0x80, 0x10, 0x09, 0x78,
+	0x0a, 0x68, 0x37, 0xa4, 0x54, 0x48, 0x9a, 0x25,
+	0x0f, 0x4a, 0x4c, 0x21, 0x08, 0x00, 0x01, 0x01,
 	0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
-	0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x02, 0x3a,
-	0x80, 0x18, 0x71, 0x38, 0x2d, 0x40, 0x58, 0x2c,
-	0x45, 0x00, 0xf4, 0x19, 0x11, 0x00, 0x00, 0x1e,
-	0x00, 0x00, 0x00, 0xff, 0x00, 0x4c, 0x69, 0x6e,
-	0x75, 0x78, 0x20, 0x23, 0x30, 0x0a, 0x20, 0x20,
-	0x20, 0x20, 0x00, 0x00, 0x00, 0xfd, 0x00, 0x3b,
-	0x3d, 0x42, 0x44, 0x0f, 0x00, 0x0a, 0x20, 0x20,
-	0x20, 0x20, 0x20, 0x20, 0x00, 0x00, 0x00, 0xfc,
-	0x00, 0x4c, 0x69, 0x6e, 0x75, 0x78, 0x20, 0x46,
-	0x48, 0x44, 0x0a, 0x20, 0x20, 0x20, 0x00, 0x05,
+	0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x1d,
+	0x80, 0x18, 0x71, 0x1c, 0x16, 0x20, 0x58, 0x2c,
+	0x25, 0x00, 0x10, 0x09, 0x00, 0x00, 0x00, 0x9e,
+	0x01, 0x1d, 0x00, 0x72, 0x51, 0xd0, 0x1e, 0x20,
+	0x6e, 0x28, 0x55, 0x00, 0x10, 0x09, 0x00, 0x00,
+	0x00, 0x1e, 0x00, 0x00, 0x00, 0xfc, 0x00, 0x34,
+	0x32, 0x20, 0x46, 0x48, 0x44, 0x5f, 0x4c, 0x43,
+	0x44, 0x2d, 0x54, 0x56, 0x00, 0x00, 0x00, 0xfd,
+	0x00, 0x31, 0x3d, 0x0f, 0x2e, 0x08, 0x00, 0x0a,
+	0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x01, 0x4e,
+	
+	0x02, 0x03, 0x2b, 0x71, 0x50, 0x01, 0x02, 0x85,
+	0x06, 0x11, 0x14, 0x15, 0x03, 0x07, 0x12, 0x16,
+	0x1f, 0x10, 0x13, 0x04, 0x20, 0x26, 0x09, 0x07,
+	0x07, 0x15, 0x07, 0x50, 0x83, 0x01, 0x00, 0x00,
+	0x6a, 0x03, 0x0c, 0x00, 0x10, 0x00, 0x88, 0x1e,
+	0x00, 0x00, 0x00, 0x8c, 0x0a, 0xd0, 0x8a, 0x20,
+	0xe0, 0x2d, 0x10, 0x10, 0x3e, 0x96, 0x00, 0x04,
+	0x03, 0x00, 0x00, 0x00, 0x18, 0x01, 0x1d, 0x80,
+	0x18, 0x71, 0x1c, 0x16, 0x20, 0x58, 0x2c, 0x25,
+	0x00, 0x10, 0x09, 0x00, 0x00, 0x00, 0x9e, 0x8c,
+	0x0a, 0xa0, 0x14, 0x51, 0xf0, 0x16, 0x00, 0x26,
+	0x7c, 0x43, 0x00, 0x04, 0x03, 0x00, 0x00, 0x00,
+	0x98, 0x8c, 0x0a, 0xd0, 0x90, 0x20, 0x40, 0x31,
+	0x20, 0x0c, 0x40, 0x55, 0x00, 0x04, 0x03, 0x00,
+	0x00, 0x00, 0x18, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xbe,
 	},
 };
 
-- 
2.17.0

